### DISCUSSION TITLE
Adding kustomization.yaml causes Flux to uninstall itself

### DISCUSSION THREAD

**Discussion by mrobertsengineering** [fluxcd/flux2#3787] [Q&A] [answered]
Hello all

I'd like a sanity check please on a repo-structure that had been working okay in labs but recently seems to not.

I'm using a mono-repo approach and  installing Flux using the bootstrap command with `--path=clusters/<environment>`.  Elsewhere in the repo I have an `apps` directory with various  Kustomize configurations to apply to a cluster.

I would then add a `kustomisation.yaml` file in `clusters/<env>` and use this as the root for things to be applied to the cluster, giving me a single location to add/remove things which felt quite tidy.

Example:
![image](https://user-images.githubusercontent.com/36993847/231137819-2a4dc77d-4e06-4f45-af22-3327f13de7c0.png)

This seemed to be working fine as I played around with things a while back, but typically as I'm investigating this for 'proper' use I'm having a big problem:  as soon as the kustomisation.yaml is present in `clusters/<env>` Flux promptly uninstalls most of its controllers.

If I run `kubectl describe kustomization flux-system -n flux-system` in the cluster prior to adding the file I can see all of Flux's own components listed.  Once I add the kustomization and re-run the command it appears to consider these no longer required and starts to remove them.  Of course because it's using it's own controllers to remove itself it only gets part way before erroring. It's like Flux considers it the new root config and promptly uninstalls most of its controllers because i've not referenced them.

What is strange is that one of my lab clusters is working fine and when I run the describe command on it, Flux's components are not present, yet still seem to function.

Currently, the fix seems to be to add the `flux-system` directory to the kustomization  to ensure that Flux understands to keep this.  What I find odd is that It feels like this would cause a self-referencing loop?
![image](https://user-images.githubusercontent.com/36993847/231142740-96ffc4dd-8936-487c-8701-095d906c28fc.png)

This is what I _thought_ Flux was doing to apply to the cluster:
![image](https://user-images.githubusercontent.com/36993847/231139708-bfd22cbc-04a8-4c7e-bddc-d36f71b5c199.png)

I may well be missing something obvious here about how Flux works through these files but happy to be corrected :)

thank you

**Comment by mgazza** [comment]
I've been along time user of flux, and having not used it for the past 18 months came back for another use case... and have discovered this exact scenario!

**Comment by stefanprodan** [comment, marked as answer]
Adding a kustomization.yaml to a cluster dir without including the `flux-system` overlay in the resources, is the equivalent of running `sudo rm -rf /*`. 

I suggest switching to [Flux Operator](https://github.com/controlplaneio-fluxcd/flux-operator) which protects you from this.

**Reply by mgazza** [reply]
Thank you for the prompt reply @stefanprodan on point as always :). that was the missing link!

I forgot that when there's no kustomization.yaml found flux just applies everything under the repo, which means it would cover the flux-system directory. If you then create a kustomization and miss out the flux-system directory then as @stefanprodan points out. Its gone!

**Reply by Nicola-Sergio** [reply]
In which way Flux operator protects us from this?

**Reply by stefanprodan** [reply]
The Flux manifests are no longer in Git so you cannot make Flux nuke itself

**Reply by Nicola-Sergio** [reply]
Ok, so gotk-components.yaml, gotk-sync.yaml and kustomization.yaml where where will be consumed?

**Reply by stealthybox** [reply]
Flux Operator gets the Flux controller manifests from an OCI artifact.
This is configured via a FluxInstance custom resource:
https://fluxcd.control-plane.io/operator/install/

Sync can also be configured on the same resource:
https://fluxcd.control-plane.io/operator/flux-sync/

FluxInstance can be directly applied to a cluster.

You can optionally choose to manage this custom resource from a control repo -- if so, you could still accidentally delete it, but it's more obvious what's happening since it's just a single resource in one file.

**Reply by Nicola-Sergio** [reply]
In the classic approach, however, if I were to find myself faced with a Flux installation that has self-uninstalled, what is the approach I should follow to correctly restore the environment?

**Reply by Nicola-Sergio** [reply]
@stealthybox But with **FluxInstance**, if I put another kustomization.yaml in the GitRepo synchronized, I could obtain the same error?

**Reply by mgazza** [reply]
> In the classic approach, however, if I were to find myself faced with a Flux installation that has self-uninstalled, what is the approach I should follow to correctly restore the environment?

I just re applied my terraform and it came back( I had to run twice because of crds)

**Reply by Nicola-Sergio** [reply]
> > In the classic approach, however, if I were to find myself faced with a Flux installation that has self-uninstalled, what is the approach I should follow to correctly restore the environment?
> 
> I just re applied my terraform and it came back( I had to run twice because of crds)

Did you have to manually delete the crds first and then reapply terraform?

**Reply by mgazza** [reply]
Nope just ran the TF. Killed the process as it hangs because flux can't start because of the missing cluster roll bindings. Then run again and it just worked

**Reply by Nicola-Sergio** [reply]
> @stealthybox But with **FluxInstance**, if I put another kustomization.yaml in the GitRepo synchronized, I could obtain the same error?

Does anyone know the answer to this question?

**Reply by Nicola-Sergio** [reply]
> Nope just ran the TF. Killed the process as it hangs because flux can't start because of the missing cluster roll bindings. Then run again and it just worked

I don't understand. I usually manage the TF code via azdo pipeline, but if for some reason the pipeline fails, relaunching 2 or more times, I can't understand how it can solve the problem.
